notebooks/Generic/demo_units_custom.ipynb

```diff
@@ -0,0 +1,334 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from astropy.wcs import WCS\n",
+    "import glue_jupyter as gj\n",
+    "from glue.core import Data\n",
+    "from glue.plugins.wcs_autolinking.wcs_autolinking import WCSLink"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set up two datasets with x-axes that use different spectral coordinates (which get auto-converted because of the WCSLink) and y-axes that use different temperature units (just to show how to add custom equivalencies)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "wcs1 = WCS(naxis=1)\n",
+    "wcs1.wcs.ctype = ['WAVE']\n",
+    "wcs1.wcs.crval = [10]\n",
+    "wcs1.wcs.cdelt = [3]\n",
+    "wcs1.wcs.crpix = [1]\n",
+    "wcs1.wcs.cunit = ['cm']\n",
+    "\n",
+    "d1 = Data(temp1=np.random.random(10))\n",
+    "d1.get_component('temp1').units = 'deg_C'\n",
+    "d1.coords = wcs1"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "wcs1 = WCS(naxis=1)\n",
+    "wcs1.wcs.ctype = ['FREQ']\n",
+    "wcs1.wcs.crval = [1000]\n",
+    "wcs1.wcs.cdelt = [1000]\n",
+    "wcs1.wcs.crpix = [1]\n",
+    "wcs1.wcs.cunit = ['MHz']\n",
+    "\n",
+    "d2 = Data(temp2=np.random.random(10) + 273)\n",
+    "d2.get_component('temp2').units = 'K'\n",
+    "d2.coords = wcs1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set up a custom unit conversion class with equivalencies:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from astropy import units as u\n",
+    "from glue.core.units import unit_converter\n",
+    "\n",
+    "@unit_converter('custom-astropy')\n",
+    "class UnitConverterWithTemperature:\n",
+    "\n",
+    "    def equivalent_units(self, data, cid, units):\n",
+    "        equivalencies = u.temperature() if 'temp' in cid.label.lower() else None\n",
+    "        return map(u.Unit(units).find_equivalent_units(include_prefix_units=True, equivalencies=equivalencies), str)\n",
+    "\n",
+    "    def to_unit(self, data, cid, values, original_units, target_units):\n",
+    "        equivalencies = u.temperature() if 'temp' in cid.label.lower() else None\n",
+    "        return (values * u.Unit(original_units)).to_value(target_units, equivalencies=equivalencies)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Use the converter class defined above"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from glue.config import settings\n",
+    "settings.UNIT_CONVERTER = 'custom-astropy'"
```

Comment:
would there ever be a use-case where you'd want to use different converters for different app/viewer instances simultaneously?  If so, being able to pass the converter (or its name) to a lower-level API to override the package-level setting might be useful.

Reply:
Won't that be super confusing though in terms of reproducible science, unless there is a way to track which converter was used to convert what?

Reply:
I don't think so, it would still either be explicitly passed or defined by the viewer itself.  I'm thinking of cubeviz/mosviz where we have a spectral viewer and image viewer in the same glue app and we probably only want one of those 4 axes to use spectral equivalancies.

Reply:
As far as `astropy` is concerned, it doesn't hurt to enable equivalencies even if you never uses it, so in theory you can enable the same set of equivalencies for all the viz. The more complicated part is when you have to pass inputs to some of them but even then maybe we can make use of helper functions like `synphot`, maybe.

```python
>>> from astropy import units as u
>>> x = 1 * u.m
>>> x.to(u.cm)
<Quantity 100. cm>
>>> x.to(u.cm, u.spectral())
<Quantity 100. cm>
>>> x.to(u.cm, u.spectral_density(x))
<Quantity 100. cm>
```

Reply:
But you probably don't want to expose frequency units as choices on a non-spectral length axes (and might even want to raise an error if the user requests display units in frequency).

Reply:
I cannot think of an example in Imviz where that concern would be an issue. @kecnry , what is the use case that you worry about?

Reply:
The API of the converter gives access to the component ID during the conversion so you can always decide what equivalencies to include based on that? (See simplistic example in the docs in the glue-core PR)

Reply:
This one?

```python
    from astropy import units as u
    from glue.core.units import unit_converter

    @unit_converter('example-2')
    class ExampleUnitConverter:

        def equivalent_units(self, data, cid, units):
            equivalencies = u.temperature() if 'temp' in cid.label.lower() else None
            return map(u.Unit(units).find_equivalent_units(include_prefix_units=True, equivalencies=equivalencies), str)

        def to_unit(self, data, cid, values, original_units, target_units):
            equivalencies = u.temperature() if 'temp' in cid.label.lower() else None
            return (values * u.Unit(original_units)).to_value(target_units, equivalencies=equivalencies)
```

Reply:
Yes - in practice you wouldn't really want to use just the name of the label, but you could imagine using e.g. info from the header, or the WCS, to make decisions about what equivalencies to use.